File: src/flourish/analysis/talents/direct_spells.py
from flourish.analysis.attributor import TalentAttributor
from flourish.analysis.talents.direct_spell import DirectSpellAttributor


class EverbloomAttributor(DirectSpellAttributor):
    name = "Everbloom"
    talent_node_id = 110424
    spell_ids = {1244341}


class EfflorescenceAttributor(DirectSpellAttributor):
    name = "Efflorescence"
    talent_node_id = 82057
    spell_ids = {81269}


class VerdancyAttributor(DirectSpellAttributor):
    name = "Verdancy"
    talent_node_id = 82059
    spell_ids = {392329}


class NaturesBountyAttributor(DirectSpellAttributor):
    name = "Nature's Bounty"
    talent_node_id = 82072
    spell_ids = {1264376}


class RegenerativeHeartwoodAttributor(DirectSpellAttributor):
    name = "Regenerative Heartwood"
    talent_node_id = 82075
    spell_ids = {392117}


class CultivationAttributor(DirectSpellAttributor):
    name = "Cultivation"
    talent_node_id = 82056
    spell_ids = {200390}


class YserasGiftAttributor(DirectSpellAttributor):
    name = "Ysera's Gift"
    talent_node_id = 82048
    spell_ids = {145108, 145109, 145110}


class EmbraceOfTheDreamAttributor(DirectSpellAttributor):
    name = "Embrace of the Dream"
    talent_node_id = 82070
    spell_ids = {392124}


class RampantGrowthAttributor(TalentAttributor):
    """Rampant Growth: +100% Regrowth HoT healing.
    Credits the bonus portion of periodic ticks only (direct heals unaffected).
    The extra Regrowth on the LB target uses spell 8936 in WCL (indistinguishable)."""
    name = "Rampant Growth"
    talent_node_id = 82058
    REGROWTH = 8936
    MULTIPLIER = 1.0  # +100%

    def process_heal(self, event, hot_tracker, buff_tracker) -> float:
        if event.ability_id == self.REGROWTH and event.tick:
            return event.amount - event.amount / (1 + self.MULTIPLIER)
        return 0.0



class FlourishAttributor(DirectSpellAttributor):
    """Flourish: Tranquility extends HoTs. The extended Tranq ticks
    appear under a different spell ID."""
    name = "Flourish"
    talent_node_id = 82053
    talent_id = 103106  # WCL entryId (choice node vs Inner Peace)
    spell_ids = {1264659}  # Tranquility (Flourish-modified)


class ThrivingVegetationAttributor(DirectSpellAttributor):
    """Thriving Vegetation: Rejuvenation instantly heals for 15/30% of its total periodic effect."""
    name = "Thriving Vegetation"
    talent_node_id = 82068
    spell_ids = {447132}

File: go-backend/internal/talents/direct_spells.go
package talents

import (
	"github.com/rdruid-talent-analyzer/go-backend/internal/models"
	"github.com/rdruid-talent-analyzer/go-backend/internal/tracking"
)

// DirectSpellAttributor attributes all effective healing from specific spell IDs.
type DirectSpellAttributor struct {
	BaseAttributor
	SpellIDs       map[int]bool
	AllowPetSource bool
}

func newDirectSpell(name string, nodeID *int, talentID *int, spellIDs []int) *DirectSpellAttributor {
	m := make(map[int]bool)
	for _, id := range spellIDs {
		m[id] = true
	}
	return &DirectSpellAttributor{
		BaseAttributor: NewBaseAttributor(name, nodeID, talentID),
		SpellIDs:       m,
	}
}

func (a *DirectSpellAttributor) ProcessHeal(event *models.HealEvent, hot *tracking.HotTracker, buff *tracking.BuffTracker) float64 {
	if a.SpellIDs[event.AbilityID] {
		if !a.AllowPetSource && a.IsPlayerPet(event.GetBase().SourceID) {
			return 0.0
		}
		return float64(event.Amount)
	}
	return 0.0
}

func intPtr(v int) *int { return &v }

func NewEverbloomSplashAttributor() *DirectSpellAttributor {
	return newDirectSpell("Everbloom: Splash", intPtr(110424), intPtr(137039), []int{EverbloomSplash})
}

func NewEfflorescenceAttributor() *DirectSpellAttributor {
	return newDirectSpell("Efflorescence", intPtr(82057), nil, []int{Efflorescence})
}

func NewVerdancyAttributor() *DirectSpellAttributor {
	return newDirectSpell("Verdancy", intPtr(82059), nil, []int{Verdancy})
}

// GroveGuardiansAttributor attributes treant pet healing, dividing out
// Wild Synthesis and Bounteous Bloom multipliers to avoid double-counting.
type GroveGuardiansAttributor struct {
	DirectSpellAttributor
	divisor float64
}

func NewGroveGuardiansAttributor() *GroveGuardiansAttributor {
	base := newDirectSpell("Grove Guardians", intPtr(82043), nil, []int{GroveGuardianNourish, GroveGuardianHeal})
	base.AllowPetSource = true
	return &GroveGuardiansAttributor{DirectSpellAttributor: *base, divisor: 1.0}
}

const (
	wildSynthesisNode      = 94535
	bounteousBloomNode     = 94591
	bounteousBloomEntry    = 117184
	wildSynthesisMultDivisor = 1.3 // 1 + 0.3 (Wild Synthesis / Bounteous Bloom multiplier)
)

func (a *GroveGuardiansAttributor) SetCombatantInfo(info *models.CombatantInfoEvent) {
	a.BaseAttributor.SetCombatantInfo(info)
	a.divisor = 1.0
	if a.HasTalent(wildSynthesisNode) {
		a.divisor *= wildSynthesisMultDivisor
	}
	if info != nil && info.TalentNodes[bounteousBloomNode] && info.TalentIDs[bounteousBloomEntry] {
		a.divisor *= wildSynthesisMultDivisor
	}
}

func (a *GroveGuardiansAttributor) ProcessHeal(event *models.HealEvent, hot *tracking.HotTracker, buff *tracking.BuffTracker) float64 {
	base := a.DirectSpellAttributor.ProcessHeal(event, hot, buff)
	if base > 0 && a.divisor > 1.0 {
		return base / a.divisor
	}
	return base
}

func NewDreamSurgeAttributor() *DirectSpellAttributor {
	return newDirectSpell("Dream Surge", intPtr(94600), nil, []int{DreamSurge})
}

func NewSpiritOfTheThicketAttributor() *DirectSpellAttributor {
	a := newDirectSpell("Spirit of the Thicket", intPtr(109712), nil, []int{SpiritOfTheThicket})
	a.AllowPetSource = true
	return a
}

func NewBurstingGrowthAttributor() *DirectSpellAttributor {
	return newDirectSpell("Bursting Growth", intPtr(109716), nil, []int{BurstingGrowthSpell})
}

func NewThrivingGrowthAttributor() *DirectSpellAttributor {
	return newDirectSpell("Thriving Growth", intPtr(94626), nil, []int{ThrivingGrowthSpell})
}

func NewNaturesBountyAttributor() *DirectSpellAttributor {
	return newDirectSpell("Nature's Bounty", intPtr(82072), nil, []int{NaturesBountySpell})
}

func NewRegenerativeHeartwoodAttributor() *DirectSpellAttributor {
	return newDirectSpell("Regenerative Heartwood", intPtr(82075), nil, []int{RegenerativeHeartwood})
}

func NewYserasGiftAttributor() *DirectSpellAttributor {
	return newDirectSpell("Ysera's Gift", intPtr(82055), nil, []int{YserasGift1, YserasGift2, YserasGift3})
}

func NewEmbraceOfTheDreamAttributor() *DirectSpellAttributor {
	return newDirectSpell("Embrace of the Dream", intPtr(82071), nil, []int{EmbraceOfTheDream})
}

func NewThrivingVegetationAttributor() *DirectSpellAttributor {
	return newDirectSpell("Thriving Vegetation", intPtr(103873), nil, []int{ThrivingVegetation})
}

func NewCultivationAttributor() *DirectSpellAttributor {
	return newDirectSpell("Cultivation", intPtr(82056), nil, []int{CultivationSpell})
}

// RampantGrowthAttributor credits bonus portion of Regrowth HoT ticks (+100%).
type RampantGrowthAttributor struct {
	BaseAttributor
}

func NewRampantGrowthAttributor() *RampantGrowthAttributor {
	return &RampantGrowthAttributor{
		BaseAttributor: NewBaseAttributor("Rampant Growth", intPtr(82058), nil),
	}
}

func (a *RampantGrowthAttributor) ProcessHeal(event *models.HealEvent, hot *tracking.HotTracker, buff *tracking.BuffTracker) float64 {
	if event.AbilityID == Regrowth && event.Tick {
		return float64(event.Amount) - float64(event.Amount)/2.0
	}
	return 0.0
}
